LFS-1022 / LFS-1060: Remove Progeny/Ancestors and wrong parent selected when creating new tumor

OrdiNeu opened the pull request:
Removes the `ancestors` and `progeny` property DefaultEditors.

[Jira link and rationale](https://phenotips.atlassian.net/browse/LFS-1022)

**To test:** Ensure that nothing involving parentage has broken, including creating tumor regions of tumors of parents, etc.

OrdiNeu committed:
LFS-1022: Remove Progeny/Ancestors

sdumitriu commented on modules/subjects/src/main/java/ca/sickkids/ccm/lfs/subjects/internal/SubjectParentEditor.java:
```diff
@@ -72,54 +70,48 @@ public SubjectAncestryEditor(final NodeBuilder nodeBuilder, final Session sessio
 
     /**
-     * Gather all parents of a given node and store them in the {@code parents} property.
+     * Gather the parent of a given node and store it in the {@code parents} property, removing it if there is none.
```
"gather one" doesn't sound right.

OrdiNeu replied:
Fixed in `acb3393` :heavy_check_mark:

sashaandjic commented:
BUG: I can create multiple tumors and tumor regions of the same name for the same patient

![image](https://user-images.githubusercontent.com/23243244/111540521-9fe4d780-8745-11eb-9b8a-7730902f1d71.png)
![image](https://user-images.githubusercontent.com/23243244/111540562-ac693000-8745-11eb-973d-c7a748ec0fac.png)
![image](https://user-images.githubusercontent.com/23243244/111540629-c276f080-8745-11eb-918e-43f24a55d75e.png)

OrdiNeu committed:
LFS-1022: Remove Progeny/Ancestors

Docstring fix.

OrdiNeu committed:
LFS-1022: Remove Progeny/Ancestors

Fix a bug that allowed multiple children with the same identifier from
being created.

OrdiNeu commented:
@sashaandjic 
> BUG: I can create multiple tumors and tumor regions of the same name for the same patient 

Should be fixed in `67128c7` :heavy_check_mark:

sashaandjic commented:
Steps to reproduce:
- create tumor region **`abc/lung/slice4`** ( patient `abc`; tumor `lung`; tumor region `slice4 `)
- create another tumor region **`slice4`**
- try to assign it to **`abc/lung`** tumor
- error msg pops up "TumorRegion slice4 already exists for Tumor lung" as expected
- click on `CONTINUE` bttn 
- click on `NEW SUBJECT` bttn
- create new tumor **`mouth`**
- the same msg pops up "TumorRegion slice4 already exists for Tumor lung."
![image](https://user-images.githubusercontent.com/23243244/112589940-572dbe00-8dd8-11eb-988f-d64a2705ec77.png)

_Please note the the error msg needs better wording._

marta- commented:
> Steps to reproduce:
> 
> * create tumor region **`abc/lung/slice4`** ( patient `abc`; tumor `lung`; tumor region `slice4 `)
> * create another tumor region **`slice4`**
> * try to assign it to **`abc/lung`** tumor
> * error msg pops up "TumorRegion slice4 already exists for Tumor lung" as expected
> * click on `CONTINUE` bttn
> * click on `NEW SUBJECT` bttn
> * create new tumor **`mouth`**
> * the same msg pops up "TumorRegion slice4 already exists for Tumor lung."

I confirm the bug.
@OrdiNeu can we gray out proposed parent subjects that already have a child with the new id? IMO the user should never have been allowed to press on CONTINUE at step 3.

OrdiNeu committed:
LFS-1060: Wrong parent selected when creating new tumor

OrdiNeu commented:
@sashaandjic 
> Steps to reproduce:
> 
>     * create tumor region **`abc/lung/slice4`** ( patient `abc`; tumor `lung`; tumor region `slice4 `)

Fixed in `1ebe85e` :heavy_check_mark:

OrdiNeu committed:
LFS-1060: Wrong parent selected when creating new tumor

Gray out ineligble parents when creating new children.

OrdiNeu commented:
@marta- 

> @OrdiNeu can we gray out proposed parent subjects that already have a child with the new id? IMO the user should never have been allowed to press on CONTINUE at step 3. E.g. in the screenshot below, "Patient abc / Tumor lung" should look disabled, clicking it should not activate the CONTINUE button and should pop the error message "Tumor region slice4 already exists for Tumor lung" (that is currently shown at the next step, see 2nd screenshot). 

Fixed in `981f201` :heavy_check_mark:

OrdiNeu committed:
LFS-1060: Wrong parent selected when creating new tumor

Fix the error message when creating duplicate children.

OrdiNeu commented:
> > Steps to reproduce:
> > 
> > * create tumor region **`abc/lung/slice4`** ( patient `abc`; tumor `lung`; tumor region `slice4 `)
> > * create another tumor region **`slice4`**
> > * try to assign it to **`abc/lung`** tumor
> > * error msg pops up "TumorRegion slice4 already exists for Tumor lung" as expected
> > * click on `CONTINUE` bttn
> > * click on `NEW SUBJECT` bttn
> > * create new tumor **`mouth`**
> > * the same msg pops up "TumorRegion slice4 already exists for Tumor lung."
> 
> I confirm the bug.
> 
> @OrdiNeu can we gray out proposed parent subjects that already have a child with the new id? IMO the user should never have been allowed to press on CONTINUE at step 3. E.g. in the screenshot below, "Patient abc / Tumor lung" should look disabled, clicking it should not activate the CONTINUE button and should pop the error message "Tumor region slice4 already exists for Tumor lung" (that is currently shown at the next step, see 2nd screenshot).
> 
> ![image](https://user-images.githubusercontent.com/651980/112680019-3e0e2700-8e43-11eb-93fa-437d3af000b5.png)
> 
> ![image](https://user-images.githubusercontent.com/651980/112680328-9fce9100-8e43-11eb-9020-1fbe75f26b1d.png)
> 
> Probably a bit out of scope here, but we should always use the SubjectType's label, not @name (e.g. "Tumor region", not "TumorRegion" as can be seen in the 2nd screenshot).

![image](https://user-images.githubusercontent.com/4656440/112692388-3b1c3200-8e55-11eb-9955-1529647bd951.png)
Fixed in `573c573` :heavy_check_mark:

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -234,13 +238,36 @@ private String fullTextEscape(String input)
         ArrayList<JsonObject> list = new ArrayList<>();
         while (resources.hasNext()) {
-            list.add(resources.next().adaptTo(JsonObject.class));
+            Resource resource = resources.next();
```
I think you can do this in one line with something like: `resourceResolver.resolve(resources.next() + ".deep").adaptTo(JsonObject.class)`

OrdiNeu replied:
I think that'd require resources to be a path instead. I could go `resources.next().getPath() + ".deep"`, but then I'd need to pass down a ResourceResolver and do a lookup every time.

sashaandjic commented:
I can still click on the grayed out option. 
![image](https://user-images.githubusercontent.com/23243244/113176825-e55fd500-921a-11eb-87c2-883a429f8eb8.png)
after which I get:
![image](https://user-images.githubusercontent.com/23243244/113176895-f6104b00-921a-11eb-8d4c-851a5f57dcb0.png)

OrdiNeu committed:
LFS-1060: Wrong parent selected when creating new tumor

Fix an issue where the greyed out option was still selectable.

OrdiNeu committed:
LFS-1060: Wrong parent selected when creating new tumor

Apply code review suggestions by @marta-

OrdiNeu commented:
@marta- 
>       Suggestion: `hasChildWithId(nodeData, childId)`

Fixed in `f749564` :heavy_check_mark:

sashaandjic reviewed (approved):
tested
